### PR TITLE
Add missing `Context::find()` method

### LINKED ISSUE #445: Missing `Context::Iterator` functions
In GitLab by @fxmordant on Oct 25, 2023, 10:56

PDI's `Context` class provides a way to iterate on `Data_descriptor`s using the `Iterator` class. Several API plugin functions are available to the developer, such as `Context::begin()` and `Context::end()`.

However, the `Context::find()` function is not available and would be welcome.

### PR DESCRIPTION
**Pull request by benoitmartin88**
Fix #445

PDI's Context class can now provide a very useful find() method. 

# List of things to check before making a PR

Before merging your code, please check the following:

* [x] you have added a line describing your changes to the Changelog;
* [x] you have added unit tests for any new or improved feature;
* [x] In case you updated dependencies, you have checked pdi/docs/CheckList.md
* you have checked your code format:
  - [x] you have checked that you respect all conventions specified in CONTRIBUTING.md;
  - [x] you have checked that the indentation and formatting conforms to the `.clang-format`;
  - [x] you have documented with doxygen any new or changed function / class;
* you have correctly updated the copyright headers:
  - [x] your institution is in the copyright header of every file you (substantially) modified;
  - [x] you have checked that the end-year of the copyright there is the current one;
* you have updated the AUTHORS file:
  - [x] you have added yourself to the AUTHORS file;
  - [x] if this is a new contribution, you have added it to the AUTHORS file;
* you have added everything to the user documentation:
  - [x] any new CMake configuration option;
  - [x] any change in the yaml config;
  - [x] any change to the public or plugin API;
  - [x] any other new or changed user-facing feature;
  - [x] any change to the dependencies;
* you have correctly linked your MR to one or more issues:
  - [x] your MR solves an identified issue;
  - [x] your commit contain the `Fix #issue` keyword to autoclose the issue when merged.

### CODE REVIEW

File: pdi/include/pdi/ref_any.h [hunk 1]
```diff
@@ -167,13 +167,13 @@ class PDI_EXPORT Reference_base
...
-				readable,
-				writable
-			))
+				  [data, freefunc, type]() {
+					  type->destroy_data(data);
+					  freefunc(data);
+				  },
+				  readable,
+				  writable
+			  ))
```
Comment (jbigot) [inline pdi/include/pdi/ref_any.h:176]: Why the change?

File: pdi/src/global_context.h [hunk 2]
```diff
@@ -108,6 +108,10 @@ class PDI_EXPORT Global_context: public Context
 	 */
 	Iterator end() override;
 
+	/** Returns the iterator that corresponds to the provided Data_descriptor name
+	 */
+	Iterator find(const std::string& name);
```
Comment (jbigot) [inline pdi/src/global_context.h:113]:
```suggestion
	Iterator find(const std::string& name) override;
```

File: pdi/tests/PDI_context.cxx [hunk 3]
```diff
@@ -217,3 +218,31 @@ TEST_F(ContextTest, iterator)
...
+
+	// Iterating through the descriptors to ensure we can find() them
+	for (set<string>::iterator it = desc_names.begin(); it != desc_names.end(); ++it) {
+		Context::Iterator descriptor = this->test_context->find(*it);
+		ASSERT_EQ(descriptor->name(), (*it));
+		ASSERT_TRUE(descriptor != this->test_context->end());
+	}
+
+	// test case where search key is not found
+	ASSERT_FALSE(this->test_context->find("desc4") != this->test_context->end()); // TODO: add operator== to Context::Iterator
```
Comment (jbigot) [inline pdi/tests/PDI_context.cxx:247]: Can you make the todo into an issue instead?

File: pdi/tests/expression.cxx [hunk 4]
```diff
@@ -78,22 +78,8 @@ TYPED_TEST(ExpressionU@CASE_NAME@Test, toLong)
...
-			EXPECT_CALL(o2_desc, ref())
-				.WillOnce(testing::Return(Ref(
-					&op2,
-					[](void*) {},
-					Scalar_datatype::type_for_v<decltype(op2)>,
-					true,
-					false
-				)));
+			EXPECT_CALL(o1_desc, ref()).WillOnce(testing::Return(Ref(&op1, [](void*) {}, Scalar_datatype::type_for_v<decltype(op1)>, true, false)));
+			EXPECT_CALL(o2_desc, ref()).WillOnce(testing::Return(Ref(&op2, [](void*) {}, Scalar_datatype::type_for_v<decltype(op2)>, true, false)));
```
Comment (jbigot) [inline pdi/tests/expression.cxx:82]: Why the change?

File: pdi/tests/expression.cxx [hunk 5]
```diff
@@ -119,22 +105,8 @@ TYPED_TEST(ExpressionU@CASE_NAME@Test, toDouble)
...
```
Comment (jbigot) [inline pdi/tests/expression.cxx:109]: Why the change?

File: pdi/tests/expression.cxx [hunk 6]
```diff
@@ -160,22 +132,8 @@ TYPED_TEST(ExpressionU@CASE_NAME@Test, toRef)
...
```
Comment (jbigot) [inline pdi/tests/expression.cxx:136]: Why the change?

File: pdi/tests/modulo.cxx [hunk 7]
```diff
@@ -65,22 +65,8 @@ TEST(ModuloU@CASE_NAME@Test, toLong)
...
-			EXPECT_CALL(o2_desc, ref())
-				.WillOnce(testing::Return(Ref(
-					&op2,
-					[](void*) {},
-					Scalar_datatype::type_for_v<decltype(op2)>,
-					true,
-					false
-				)));
+			EXPECT_CALL(o1_desc, ref()).WillOnce(testing::Return(Ref(&op1, [](void*) {}, Scalar_datatype::type_for_v<decltype(op1)>, true, false)));
+			EXPECT_CALL(o2_desc, ref()).WillOnce(testing::Return(Ref(&op2, [](void*) {}, Scalar_datatype::type_for_v<decltype(op2)>, true, false)));
```
Comment (jbigot) [inline pdi/tests/modulo.cxx:69]: Why the change?

File: pdi/tests/modulo.cxx [hunk 8]
```diff
@@ -104,22 +90,8 @@ TEST(ModuloU@CASE_NAME@Test, toDouble)
...
```
Comment (jbigot) [inline pdi/tests/modulo.cxx:94]: Why the change?

File: pdi/tests/modulo.cxx [hunk 9]
```diff
@@ -143,22 +115,8 @@ TEST(ModuloU@CASE_NAME@Test, toRef)
...
```
Comment (jbigot) [inline pdi/tests/modulo.cxx:119]: Why the change?

File: pdi/CHANGELOG.md [hunk 10]
```diff
@@ -9,12 +9,13 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
```
Comment (jbigot) [inline pdi/CHANGELOG.md:11]:
```suggestion


```

**Review by Yushan-Wang — APPROVED** [review]
Great!

File: pdi/AUTHORS [hunk 11]
```diff
@@ -10,10 +10,14 @@ Julien Bigot - CEA (julien.bigot@cea.fr)
...
 * Initial Fortran interface
 
+Benoit Martin - CEA (bmartin@cea.fr)
+* Add Context::find()
+
 François-Xavier Mordant - CEA (francois-xavier.mordant@cea.fr)
 * Fixed CMake issues, internal API enhancement
 * Add support for ZSH in env
 * Update supported CMake versions
+* Internal API improvements
```
Comment (jbigot) [inline pdi/AUTHORS:20]:
```suggestion
* Initial work on `Context::find()`
```

File: AUTHORS [hunk 12]
```diff
@@ -11,6 +11,9 @@ Julien Bigot - CEA (julien.bigot@cea.fr)
 * Maintainer (Dec. 2014 - ...)
 * Design and initial implementation
 
+Benoit Martin - CEA (bmartin@cea.fr)
+* Maintainer (Jan. 2024 - ...)
+
```
Comment (jbigot) [inline AUTHORS:16]: I believe this would be better moved to a separate PR

File: pdi/include/pdi/context.h [hunk 13]
```diff
@@ -96,6 +96,13 @@ class PDI_EXPORT Context
 	 */
 	virtual Iterator end() = 0;
 
+	/** Find the Data_descriptor corresponding to a given name
+	 * \param[in] name the Data_descriptor name
+	 *
+   * \return an iterator to the requested Data_descriptor name. If no such element is found, past-the-end (see end()) iterator is returned.
```
Comment (jbigot) [inline pdi/include/pdi/context.h:102]:
```suggestion
	 * \return an iterator to the requested Data_descriptor name. If no such element is found, past-the-end (see end()) iterator is returned.
```

**Review by jbigot — DISMISSED** [review]
Looks good, only minor points remaining

File: pdi/include/pdi/context.h [hunk 14]
```diff
@@ -96,6 +96,13 @@ class PDI_EXPORT Context
 	 */
 	virtual Iterator end() = 0;
 
+	/** Find the Data_descriptor corresponding to a given name
+	 * \param[in] name the Data_descriptor name
+	 *
```
Comment (jbigot) [inline pdi/include/pdi/context.h:101]:
```suggestion
	 *
	 * \param[in] name the Data_descriptor name
```

**Review by jbigot — APPROVED** [review]
Looks great, thanks!

**Review by Yushan-Wang — APPROVED** [review]
